tools/schemacode/schemacode/validator.py

```diff
@@ -0,0 +1,733 @@
+
+    # Standardize Windows paths
+    if "\\" in path_list[0]:
+        for ix, i in enumerate(path_list):
+            path_list[ix] = i.replace("\\", "/")
+
+    return path_list
+
+
+def _add_entity(regex_entities, entity, entity_shorthand, variable_field, requirement_level):
+    """Add entity pattern to filename template based on requirement level."""
+
+    # We need to do this here, although it would be easier to back-reference in the directory.
+    # This is because regex evaluates sequentially and we can not forward-reference a group.
+    if entity in DIR_ENTITIES:
+        variable_regex = f"(?P={entity})"
+    else:
+        variable_regex = f"(?P<{entity}>{variable_field})"
+
+    if requirement_level == "required":
+        if len(regex_entities.strip()):
+            regex_entities += f"_{entity_shorthand}-{variable_regex}"
+        else:
+            # Only the first entity doesn't need an underscore
+            regex_entities += f"{entity_shorthand}-{variable_regex}"
+    else:
+        if len(regex_entities.strip()):
+            regex_entities += f"(|_{entity_shorthand}-{variable_regex})"
+        else:
+            # Only the first entity doesn't need an underscore
+            regex_entities += f"(|{entity_shorthand}-{variable_regex})"
+
+    return regex_entities
+
+
+def _extension_safety(extension):
+    """
+    Making extensions formatting-safe.
+    Issues covered by this function are listed under “Notes”
+
+    Parameters
+    ----------
+    extension : str
+        Extension string, as present in the BIDS YAML schema.
+
+    Returns
+    -------
+    str
+        Extension string, safe for use in validator Regex formatting.
+
+    Notes
+    -----
+    * Bash-wildcard safety: https://github.com/bids-standard/bids-specification/issues/990
+    * Period safety: https://github.com/bids-standard/bids-specification/issues/1055
+    * Hopefully this function will be deprecated soon, but it will not break safe entries.
+    """
+    if extension == "None":
+        return ""
+    if "." in extension:
+        extension = extension.replace(".", "\\.")
+    if "*" in extension:
+        extension = extension.replace("*", ".*?")
+
+    return extension
+
+
+def _add_extensions(regex_string, variant):
+    """Add extensions to a regex string."""
+    fixed_variant_extensions = []
+    for variant_extension in variant["extensions"]:
+        variant_extension = _extension_safety(variant_extension)
+        fixed_variant_extensions.append(variant_extension)
+    if len(fixed_variant_extensions) > 1:
+        regex_extensions = "({})".format("|".join(fixed_variant_extensions))
+    else:
+        regex_extensions = fixed_variant_extensions[0]
+    regex_string = f"{regex_string}{regex_extensions}"
+
+    return regex_string
+
+
+def _add_subdirs(regex_string, variant, datatype, entity_definitions, modality_datatypes):
+    """Add appropriate subdirectories as required by entities present."""
+
+    label = "([a-zA-Z0-9]*?)"
```

Comment:
I think we want at least one character, and greedy. These should always end with `_` or `.` or `/`, right? And shouldn't we have indices for things like `run` and `echo`?

```suggestion
    label = "([a-zA-Z0-9]+)"
    index = "(\d+)" 
```

Reply:
loaded it from the schema, should be the most.... schema-tic ;) way to do it.